### PR TITLE
fix(css): elaborate how gap percentage values behave

### LINKED ISSUE #24933: Interpretation of percentage values in grid-gap
### MDN URL

https://developer.mozilla.org/en-US/docs/Web/CSS/gap

### What specific section or headline is this issue about?

Syntax

### What information was incorrect, unhelpful, or incomplete?

/* One <percentage> value */
gap: 16%;
gap: 100%;

### What did you expect to see?

Percentage values at grid-gap leads to problems in calculating the height of the grid element. This happens in Firefox, Chrome, Opera and Edge (tested). 
So it should be mentioned in this manual.
Thank you

### Do you have any supporting links, references, or citations?

_No response_

### Do you have anything more you want to share?

_No response_

### MDN metadata

<details>
<summary>Page report details</summary>

* Folder: `en-us/web/css/gap`
* MDN URL: https://developer.mozilla.org/en-US/docs/Web/CSS/gap
* GitHub URL: https://github.com/mdn/content/blob/main/files/en-us/web/css/gap/index.md
* Last commit: https://github.com/mdn/content/commit/73091fbe590d96857d743eaeec5aee4a8101994f
* Document last modified: 2023-02-21T05:50:15.000Z

</details>

### PR DESCRIPTION
**Pull request by OnkarRuikar**
- fixes https://github.com/mdn/content/issues/24933

### CODE REVIEW

File: files/en-us/web/css/gap/index.md [hunk 1]
```diff
@@ -176,6 +176,28 @@ This property is specified as a value for `<'row-gap'>` followed optionally by a
...
 
+## Percentage gap values
+
+Gap acts like an empty item/track. In other words instead of gap you could simply use an empty item/track in the container. The item/track's size would be the size set in `gap` property value. Percentage values are always calculated against the [content box](/en-US/docs/Learn/CSS/Building_blocks/The_box_model#parts_of_a_box) size of the container element.
+
+### Explicit container size
+
+If the container has fixed size set then percentage value calculations are simple. Percentages are simply calculated from the sizes set on the container. Thus gap behaviour is consistant across all layouts. In the following screenshot there are two containers one with grid layout and other with flex layout. The containers have five red 20x20px children. Both the containers have height explicitly set to 200px using `height: 200px`. And the gap is set using `gap: 10% 0`.
+
+![Screenshot demoing gap sizes in grid and flex layouts with fixed 200px height set.](fixed_height.jpg)
```
Comment (bsmth) [inline files/en-us/web/css/gap/index.md:187]: Quick one before I go deeper in this review: shall we do a live sample instead of screenshots?

Comment (OnkarRuikar) [reply files/en-us/web/css/gap/index.md:187]: The screenshots can't be replicated using livesample; it is not possible to color gaps. In developer tools when you select grid/flex elements in HTML tab you get to see the sizes, all the gaps, and the highlighting.

Comment (bsmth) [reply files/en-us/web/css/gap/index.md:187]: We could do live samples and point readers where to look for layout inspector in developer tools to highlight the gaps? I would really like to avoid adding more screenshots 😂 and the code can be remixed and maintained afterwards! What do you think?

Comment (OnkarRuikar) [reply files/en-us/web/css/gap/index.md]: removed the screenshots and used live samples instead

Comment (bsmth) [reply files/en-us/web/css/gap/index.md]: Looks much better IMO, thank you! 🙏🏻

File: files/en-us/web/css/gap/index.md [hunk 2]
```diff
@@ -190,6 +190,120 @@ Early versions of the specification called this property `grid-gap`, and to main
...
+  display: inline-flex;
+  height: 200px;
+  gap: 10% 0;
+}
+
+#grid > div,
+#flex > div {
+  background-color: coral;
+  width: 20px;
+  height: 20px;
```
Comment (bsmth) [inline files/en-us/web/css/gap/index.md:245]: Note that we have five gaps in the flex layout and four in the grid layout (possibly margin collapse):

image

we could do something like this with `min-height` so boxes are 24px

https://codepen.io/bsmth/pen/qBGYwyZ

In the current example, the gap is the same but the layouts look different. I'm not sure if that's intentional. What do you think?

File: files/en-us/web/css/gap/index.md [hunk 3]
```diff
@@ -190,6 +190,120 @@ Early versions of the specification called this property `grid-gap`, and to main
 
 {{EmbedLiveSample("Multi-column_layout", "auto", "120px")}}
 
+## Percentage gap values
+
+Gap acts like an empty item/track. In other words instead of gap you could simply use an empty item/track in the container. The item/track's size would be the size set in `gap` property value. Percentage values are always calculated against the [content box](/en-US/docs/Learn/CSS/Building_blocks/The_box_model#parts_of_a_box) size of the container element.
```
Comment (bsmth) [inline files/en-us/web/css/gap/index.md:195]: Removing "simply":

```suggestion
A gap acts like an empty item or track. Instead of the gap, you could use an empty item or track in the container, and the item or track's size would be the size set in the `gap` property value. Percentage values are always calculated against the [content box](/en-US/docs/Learn/CSS/Building_blocks/The_box_model#parts_of_a_box) size of the container element.
```

File: files/en-us/web/css/gap/index.md [hunk 4]
```diff
@@ -190,6 +190,120 @@ Early versions of the specification called this property `grid-gap`, and to main
...
+
+### Explicit container size
+
+If the container has fixed size set then percentage value calculations are simple. Percentages are simply calculated from the sizes set on the container. Thus gap behaviour is consistant across all layouts. In the following example there are two containers one with grid layout and other with flex layout. The containers have five red 20x20px children. Both the containers have height explicitly set to 200px using `height: 200px`. And the gap is set using `gap: 10% 0`.
```
Comment (bsmth) [inline files/en-us/web/css/gap/index.md:199]: Minor reword:

```suggestion
If the container has a fixed size set, then gap percentage value calculations are based on the size of the container. Thus, gap behavior is consistent across all layouts. In the following example, there are two containers, one with a grid layout and the other with a flex layout. The containers have five red 20x20px children. Both containers are explicitly set to 200px using `height: 200px` and the gap is set with `gap: 10% 0`.
```

Note we're saying it's consistent but the spacing of items looks inconsistent. We should change the example so there's an equal number of gaps in each layout or explain why this may unexpectedly happen.

File: files/en-us/web/css/gap/index.md (hunk 2 above)
Comment (OnkarRuikar) [reply files/en-us/web/css/gap/index.md:245]: > Note that we have five gaps in the flex layout and four in the grid layout (possibly margin collapse):

Interesting, it's not margin collapse though. In case of grid the height gets divided equally among 5 [grid-rows](https://developer.mozilla.org/en-US/docs/Glossary/Grid_Row) then gaps *in between* the rows are subtracted from the row heights:
```
total height: 200px
grid rows without gaps: 5 * 40 = 200px
4 10% gaps in between rows + rows = 80 + 120 = 4 * 20 + (5 * 24)
```
From above calculations grid rows are of 24px, but the items are only 20px. The remaining 4px worth of grid area is remains blank:

In above screenshot, taken from chrome, it's clear that grid rows are bigger than the grid items. Empty grid area has been pointed.

In case of flex, biggest flex-item on the row decides row height:

```
total height: 200px
flex rows + remaining empty area:  100 + 100 = 5 * 20 + 100
flex rows + grid gaps + remaining area: 100 + 80 + 20 = 5 * 20 + 4 * 20 + 20
```
In short, in case of our flex box, the items are stacked upwards and some area is left behind.


> In the current example, the gap is the same but the layouts look different. I'm not sure if that's intentional. What do you think?

It's not intentional, but the grid and flex containers are different. But we shouldn't get into such details because this is demo about gap. So we should set item heights to 24px to make both containers look more similar.

File: files/en-us/web/css/gap/index.md [hunk 5]
```diff
@@ -190,6 +190,120 @@ Early versions of the specification called this property `grid-gap`, and to main
...
+  display: inline-flex;
+  gap: 10% 0;
+  flex-flow: column;
+}
+
+#grid > div,
+#flex > div {
+  background-color: coral;
+  width: 20px;
+  height: 20px;
```
Comment (OnkarRuikar) [inline files/en-us/web/css/gap/index.md:297]:
```suggestion
  width: 24px;
  height: 24px;
```

**Review by OnkarRuikar — COMMENTED** [review]
Let's make dimensions 24px to make flex and grid look same.

File: files/en-us/web/css/gap/index.md [hunk 6]
```diff
@@ -190,6 +190,120 @@ Early versions of the specification called this property `grid-gap`, and to main
...
+#flex > div {
+  background-color: coral;
+  width: 24px;
+  height: 24px;
+}
+```
+
+{{EmbedLiveSample("Implicit container size", "auto", "200px")}}
+
+In case of the grid layout, percentage gap doesn't contribute to actual height of the grid. The container's height is calculated using `0px` gap, so the actual height turns out to be 120px (24px x5). Then the actual percentage gap is calculated using the content box's height, the gap turns out to be `12px` (120px x 10%). The gap is applied just before rendering. Thus the grid remains 120px high but it overflows due to the percentage gap added later just befor rendering.
```
Comment (OnkarRuikar) [inline files/en-us/web/css/gap/index.md:303]:
```suggestion
In case of the grid layout, percentage gap doesn't contribute to actual height of the grid. The container's height is calculated using `0px` gap, so the actual height turns out to be 120px (24px x 5). Then the actual percentage gap is calculated using the content box's height, the gap turns out to be `12px` (120px x 10%). The gap is applied just before rendering. Thus the grid remains 120px high but it overflows due to the percentage gap added later just befor rendering.
```

**Comment by bsmth** [comment]
It feels a little wonky to use 24px here instead of 20px as with the others. What do you think about this:

https://codepen.io/bsmth/pen/mdYNyJb

i.e.:

```css
body > div {
  background-color: #ccc;
  width: 200px;
  height: 200px;
  flex-flow: column;
}

#grid {
  display: inline-grid;
  gap: 12.5%;
}

#flex {
  display: inline-flex;
  gap: 12.5%;
}

#grid > div,
#flex > div {
  background-color: coral;
  width: 20px;
  height: 20px;
}
```

Or with four items and some even values

```html
<span>Grid</span>
<div id="grid">
  <div>1</div>
  <div>2</div>
  <div>3</div>
  <div>4</div>
</div>
<span>Flex</span>
<div id="flex">
  <div>1</div>
  <div>2</div>
  <div>3</div>
  <div>4</div>
</div>
```

```css
body > div {
  background-color: #ccc;
  width: 200px;
  height: 200px;
  flex-flow: column;
}

#grid {
  display: inline-grid;
  gap: 20%;
}

#flex {
  display: inline-flex;
  gap: 20%;
}

#grid > div,
#flex > div {
  background-color: coral;
  width: 20px;
  height: 20px;
}
```

**Comment by OnkarRuikar** [comment]
> It feels a little wonky to use 24px here instead of 20px as with the others. What do you think about this:

Yes, 20px is easy to explain.

File: files/en-us/web/css/gap/index.md [hunk 7]
```diff
@@ -190,6 +190,120 @@ Early versions of the specification called this property `grid-gap`, and to main
 
 {{EmbedLiveSample("Multi-column_layout", "auto", "120px")}}
 
+## Percentage gap values
+
+A gap acts like an empty item or track. Instead of the gap, you could use an empty item or track in the container, and the item or track's size would be the size set in the `gap` property value. Percentage values are always calculated against the [content box](/en-US/docs/Learn/CSS/Building_blocks/The_box_model#parts_of_a_box) size of the container element.
+
+### Explicit container size
+
+If the container has a fixed size set, then gap percentage value calculations are based on the size of the container. Thus, gap behavior is consistent across all layouts. In the following example, there are two containers, one with a grid layout and the other with a flex layout. The containers have five red 24x24px children. Both containers are explicitly set to 200px using `height: 200px` and the gap is set with `gap: 12.5% 0`.
```
Comment (bsmth) [inline files/en-us/web/css/gap/index.md:199]:
```suggestion
If the container has a fixed size set, then gap percentage value calculations are based on the size of the container. Thus, gap behavior is consistent across all layouts. In the following example, there are two containers, one with a grid layout and the other with a flex layout. The containers have five red 20x20px children. Both containers are explicitly set to 200px high using `height: 200px` and the gap is set with `gap: 12.5% 0`.
```

File: files/en-us/web/css/gap/index.md [hunk 8]
```diff
@@ -190,6 +190,120 @@ Early versions of the specification called this property `grid-gap`, and to main
 
 {{EmbedLiveSample("Multi-column_layout", "auto", "120px")}}
 
+## Percentage gap values
```
Comment (bsmth) [inline files/en-us/web/css/gap/index.md:193]: Last comment: Should this be a H2? It breaks the usual page template. What might work is:

* move this para up to the bottom of the `## Description` section saying 'there are two examples that demonstrate percentage gap values with explicit and implicit container size below' linking to:
* Rename the example H3s to "Percentage gap value and explicit container size" and "Percentage gap value and implicit container size"

What do you reckon?

**Comment by bsmth** [comment]
> Yes, 20px is easy to explain.

Looks better!

**Comment by bsmth** [comment]
I'm so sorry for dragging out this review... I've read through the description again and it can do with a reword because it's already describing percentage values. This branch has a suggested edit: https://github.com/OnkarRuikar/content/compare/css_gap...bsmth:content:css_gap_description?expand=1

Then the paras have these functions:

```md
## Description

* what it is 
* how it works / looks
* use in grid
* use in flex
* use in multi-col
* percentage values
* history
```

I'm happy to merge after this!

**Comment by OnkarRuikar** [comment]
> I've read through the description again and it can do with a reword because it's already describing percentage values.

Yes, the text made sense when it was a separate section. Now in the description section, it looks repetitive.

I've done the changes.

**Review by bsmth — APPROVED** [review]
Looks super. Thanks for your patience, Onkar!
